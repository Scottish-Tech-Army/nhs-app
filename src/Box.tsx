import React, { FormEvent, useState } from "react";
import { TRAUMA_TOWER_TEMPLATE } from "./data/traumaTower";
import { useParams } from "react-router-dom";

export type FormValueType = {
  name: string;
  count: number;
};

export interface BoxProps {
  setBoxContents: (boxTemplateId: string, boxNumber: number, contents: FormValueType[]) => void;
}

function Box({ setBoxContents }: BoxProps) {

//init variables
  //destructure nested routes from useParams  
  let { boxTemplateId, boxId } = useParams();
  //init boxTemplate var finding boxTemplateId from TRAUMA_TOWER_TEMPLATE Doc
  const boxTemplate = TRAUMA_TOWER_TEMPLATE.boxes.find(
    (box) => box.boxTemplateId === boxTemplateId
  );
  //init var of mapped items in template creating obj of name and count/amount of each
  const startingArrayOfObjects: FormValueType[] =
      boxTemplate?.items.map(({ name }) => ({ name, count: 0 })) || [];
  //init- useState hook using mapped items stored as init value  
  const [formValues, setFormValues] = useState<FormValueType[]>(
    startingArrayOfObjects
  );
  //init boxNumber to zero   
    let boxNumber = 0;
  //check boxId matches string of starting digit/end digit  
    //parse boxId as Int
    //it should so continue(maybe integrate fail condition?)
  if (boxId?.match(/^\d+$/)) {
    boxNumber = Number.parseInt(boxId);
  }
  //check- no boxTemplate OR boxNumber less than 1 OR boxNumber is more than boxTemplate count/value
    //if true return null (probably should return Error)
    //if false continue
  if (!boxTemplate || boxNumber < 1 || boxNumber > boxTemplate.count) {
    return null;
  }

  //Functions for storing form values on user input- form change/submit
    //handle change- capture change to useState, set change to setFormValues
        //init- form new form values to array
        //process values to Int
        //set to form values state
  let handleChange = (index: number, target: HTMLInputElement) => {
    let newFormValues = [...formValues];

    newFormValues[index].count = Number.parseInt(target!.value);

    setFormValues(newFormValues);
  };

   //handleSubmit- pass router values and inputted state values  
        //handle browser event to keep inputted data
        //pass router values and inputted form values to setBoxContents function
  let handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    setBoxContents(boxTemplateId!, boxNumber, formValues);
  };

  return (
    <div>
      <h1>{`${boxTemplate.name} - Box ${boxId}`}</h1>

      <form onSubmit={handleSubmit}>
        <ul>
          {boxTemplate.items.map((item, index) => (
            <li key={index}>
              <label>
                {item.name}
                <input
                  name={item.name}
                  type="number"
                  max="10"
                  min="0"
                  step="1"
                  value={formValues[index].count || 0}
                  onChange={({ target }) => handleChange(index, target)}
                />
              </label>
            </li>
          ))}
        </ul>
        <div>
          <button type="submit">Submit</button>
        </div>
      </form>
    </div>
  );
}

export default Box;
