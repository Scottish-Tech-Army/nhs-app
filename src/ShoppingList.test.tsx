import React from "react";

import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import storageAreaContents from "./data/traumaTower";

import Box from "./Box";

const setBoxContents = jest.fn();

describe("ShoppingList", () => {
  it("rendered a shopping list page for empty store - ie all items shown", async () => {
    render(
      <ShoppingList
        expectedContents={storageAreaContents}
        actualContents={[]}
      />
    );
    expect(screen.getByText("Items to replace")).toBeDefined();

    checkShoppingList([
      {
        name: "Box 1",
        items: ["item11", "item22", "item33", "item44", "item55", "item66"],
      },
      {
        name: "Box 2",
        items: ["item11", "item22", "item33", "item44", "item55", "item66"],
      },
    ]);
  });

  it("rendered a shopping list page for filled store - ie no items shown", async () => {
    render(
      <ShoppingList
        expectedContents={storageAreaContents}
        actualContents={ALL_CONTENTS}
      />
    );
    expect(screen.getByText("Items to replace")).toBeDefined();

    checkShoppingList([]);
  });

  it("rendered a shopping list page for partially filled store - ie some items shown", async () => {
    render(
      <ShoppingList
        expectedContents={storageAreaContents}
        actualContents={PARTIAL_CONTENTS}
      />
    );
    expect(screen.getByText("Items to replace")).toBeDefined();

    checkShoppingList([
      { name: "Box 2", items: ["item22", "item33", "item55"] },
    ]);
  });

  it("renders correctly", () => {
    const { container } = render(
      <ShoppingList
        expectedContents={storageAreaContents}
        actualContents={PARTIAL_CONTENTS}
      />
    );

    expect(container).toMatchSnapshot();
  });

  type ExpectedBoxContents = {
    name: string;
    items: string[];
  };

  function checkShoppingList(expectedBoxes: ExpectedBoxContents[]) {
    // eslint-disable-next-line testing-library/no-node-access
    const actualBoxes = document.querySelectorAll("div.boxes")!;

    expect(actualBoxes).toHaveLength(expectedBoxes.length);
    actualBoxes.forEach((actualBox, index) => {
      const expectedBox = expectedBoxes[index];
      expect(actualBox).toHaveTextContent(expectedBox.name);

      // eslint-disable-next-line testing-library/no-node-access
      const actualItems = document.querySelectorAll("dl div")!;
      expect(actualItems).toHaveLength(expectedBox.items.length);
      actualItems.forEach((actualItem, index) => {
        expect(actualItem).toHaveTextContent(expectedBox.items[index]);
      });
    });
  }
});
