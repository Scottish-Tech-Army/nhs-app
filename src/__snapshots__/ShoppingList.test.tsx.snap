// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingList renders correctly 1`] = `
<div>
  <div
    class="shopping-list"
  >
    <header>
      <button
        aria-label="back"
        class="back"
        type="button"
      >
        <svg>
          arrow-left.svg
        </svg>
      </button>
      <h1>
        Summary
      </h1>
    </header>
    <main>
      <div
        class="scroll"
      >
        <div
          class="box"
        >
          <div>
            <h2>
              Trauma Chest Drain - Box 2
            </h2>
            <div
              class="checker"
            >
              Checked by Bob on Thu 23/6/2022 at 10:18
            </div>
          </div>
          <div
            class="items"
          >
            <div
              class="item"
            >
              1 x Sterile gloves (Medium)
            </div>
            <div
              class="item"
            >
              1 x ChloraPrep applicator
            </div>
            <div
              class="item"
            >
              2 x Lidocaine 1% (5ml / 50mg)
            </div>
            <div
              class="item"
            >
              1 x Chest drain bottle
            </div>
          </div>
        </div>
      </div>
    </main>
  </div>
</div>
`;
